### PR TITLE
Add code review tips

### LINKED ISSUE #5: Code review checklist
https://blog.fogcreek.com/effective-code-reviews-9-tips-from-a-converted-skeptic/
http://blog.fogcreek.com/increase-defect-detection-with-our-code-review-checklist-example/

Checklist

Architectural
 - Is there any redundant or duplicate code?
 - Is the code as modular as possible?
 - Are all the queries in repositories?
 - Can any of the code be replaced with library functions?
 - Is the any hardcoded string or value that is not saved in variable (preferably constant in a class)
 - Is most of the code in helpers/services preferably used via dependency injection?
 - If you use a typeless programming language, type hinting should be used or dock blocks to specify the expected data type
 - Are there any missing updates to the documentation/readme files?

Security
 - Are all data inputs checked (for the correct type, length, format, and range) and encoded?
 - Are invalid parameter values handled?

Testing
 - Is the code testable? i.e. don’t add too many or hide dependencies, unable to initialize objects, test frameworks can use methods etc.
 - Do unit tests actually test that the code is performing the intended functionality?
 - Are dependencies mocked?

Standards
 - Name your variables to say what they are for, not what data type they are
 - Give your methods a clear name for what they are going to do.

Coding
 - There is no `Created by XYZ ….` line in the code. We use git for that.

### PR DESCRIPTION
**Pull request by stocarul**
- Add tips for having your code reviewed
 - Add tips for reviewing code
 - Add a code review checklist
 - Increase heading formatting for titles
 - Closes #5 

I know that the new `Code review` chapter should be in a dedicated file, but I plan to do that in a new commit/PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,7 +52,56 @@
...
+ - Ask good questions; don't make demands.
+ - Offer alternative implementations, but assume the author already considered them.
+
+### Code review checklist
+
+This code review checklist helps the code reviewers and software developers (during self code review) to gain expertise in the code review process.
+
+#### Architectural
+
+ - Is all the code easily understood?
```
Comment (ssergiuss) [inline README.md]: I think `understandable` is better than `understood`.

File: README.md [hunk 2]
```diff
@@ -52,7 +52,56 @@
...
+ - Are all data inputs checked (for the correct type, length, format, and range) and encoded?
+ - Are invalid parameter values handled?
+ - Are all data inputs validated against security threats such as SQL injections and Cross Site Scripting (XSS)?
+
+#### Coding
+
+ - Do all the varialbes say what they are for and not what data type they have?
+ - Do all the methods have a clear name for what they are going to do?
+ - Are there no hardcoded string/integer values?
+ - Is there no `Created by XYZ` line in the code? We use git for that.
```
Comment (ssergiuss) [inline README.md]: Be respectful. It's `Git` not `git` :)).
Also I think you should rephrase the first sentence because it's not quite clear.

**Review by ssergiuss — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (stocarul) [reply README.md]: 👍

File: README.md (hunk 2 above)
Comment (stocarul) [reply README.md]: Completely agree with you. I think it is better now, but still not if it is clear enough. Please let me know what you think.

File: README.md [hunk 3]
```diff
@@ -52,7 +52,56 @@
...
+### Code review checklist
+
+This code review checklist helps the code reviewers and software developers (during self code review) to gain expertise in the code review process.
+
+#### Architectural
+
+ - Is all the code easily understandable?
+ - Is there any duplicate code?
+ - Is the code as modular as possible?
+ - Are all the queries in repositories?
```
Comment (lgnstn) [inline README.md]: Isn't this tip a little bit too specific to a particular type of applications?

File: README.md [hunk 4]
```diff
@@ -52,7 +52,56 @@
...
+ - Are all data inputs checked (for the correct type, length, format, and range) and encoded?
+ - Are invalid parameter values handled?
+ - Are all data inputs validated against security threats such as SQL injections and Cross Site Scripting (XSS)?
+
+#### Coding
+
+ - Do all the varialbes say what they are for and not what data type they have?
+ - Do all the methods have a clear name for what they are going to do?
+ - Are there no hardcoded string/integer values?
+ - Is there no code ownership comment (like `Created by X`) in the code? We use Git for that.
```
Comment (lgnstn) [inline README.md]: I would use `any` instead of `no`. The same for line 106. What do you think?

File: README.md [hunk 5]
```diff
@@ -52,7 +52,56 @@
...
+
+#### Architectural
+
+ - Is all the code easily understandable?
+ - Is there any duplicate code?
+ - Is the code as modular as possible?
+ - Are all the queries in repositories?
+ - Can any of the code be replaced with library functions?
+ - Is most of the code in helpers/services preferably used via dependency injection?
+ - Are there any missing updates to the documentation/readme files?
```
Comment (lgnstn) [inline README.md]: Should we add something about tests somewhere?

**Review by lgnstn — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (stocarul) [reply README.md]: Agree. I will change that :)

File: README.md (hunk 5 above)
Comment (stocarul) [reply README.md]: Hmm. Probably would be a good idea.

File: README.md (hunk 3 above)
Comment (stocarul) [reply README.md]: Completely agree with you here. This is quite specific. Unfortunately, I don't know how to handle this. Should we have 2 different checklists (at the bottom) for frontend and backend? Would that be enough?

Comment (lgnstn) [reply README.md]: I am not sure it is a good idea to go that deep into details. Even if you split between front and back, there are so many languages and frameworks each with their own specifics.

Comment (stocarul) [reply README.md]: Yeah ... I think you are right again. Then what about something a bit more generic: `Is all the code in the right place (queries in repositories, data manipulators in managers, etc)?`

Comment (lgnstn) [reply README.md]: I was thinking about something like this but I didn't know how to "pack" everything. Do you think it makes sense to change line 93 to also refer to these details?

Comment (stocarul) [reply README.md]: 👍

**Comment by stocarul** [comment]
@ssergiuss @lgnstn: I hope I handled all your suggestions. Please have another look at the changes.

**Comment by stocarul** [comment]
@ssergiuss @lgnstn: 🥇

File: README.md [hunk 6]
```diff
@@ -52,7 +52,56 @@
...
+ - Ask good questions; don't make demands.
+ - Offer alternative implementations, but assume the author already considered them.
+
+### Code review checklist
+
+This code review checklist helps the code reviewers and software developers (during self code review) to gain expertise in the code review process.
+
+#### Architectural
+
+ - Is all the code easily understandable?
```
Comment (lgnstn) [inline README.md]: I would use "easy to understand".

File: README.md [hunk 7]
```diff
@@ -52,7 +52,56 @@
...
+ - Is there any duplicate code?
+ - Is the code as modular as possible?
+ - Can any of the code be replaced with library functions?
+ - Is most of the code in helpers/services/managers/repositories?
+ - Are there any missing updates to the documentation/readme files?
+ - Are all the necessary tests created according to the project guidelines?
+ 
+#### Security
+
+ - Are all data inputs checked (for the correct type, length, format, and range) and encoded?
```
Comment (lgnstn) [inline README.md]: I would remove the parentheses and I am not sure about that "encoded".

**Review by lgnstn — CHANGES_REQUESTED** [review]

**Pull request merged**